### ISSUE TITLE
Bug: `no-fallthrough` option ignores functions that never return

### ISSUE THREAD

**Issue by loganek** [eslint/eslint#17222]
### Environment

Node version: v16.3.0
npm version: 8.19.2
Local ESLint version: 8.41.0
Global ESLint version: 8.51.0
Operating System: Ubuntu 20.04.6

### What parser are you using?

@typescript-eslint/parser

### What did you do?

<details>
<summary>Configuration</summary>

```
{
    "root": true,
    "parser": "@typescript-eslint/parser",
    "rules": {
        "no-fallthrough": "error"
    }
}
```
</details>

**file.ts**

```js
function th(): never {
    throw Error();
}

export function foo(v: number): void {
    switch (v) {
        case 1:
            th();
        case 2:
            console.log("OK");
            break;
    }
}
```

### What did you expect to happen?

I'm expecting the `no-fallthrough` rule to pass. Even though there's no `break` in the first `case`, it calls function that never returns. In fact, for the following code the rule doesn't report any issue:

```ts
export function foo(v: number): void {
    switch (v) {
        case 1:
            throw Error();
        case 2:
            console.log("OK");
            break;
    }
}
```

### What actually happened?

Getting a following error:
```
file.ts
  9:9  error  Expected a 'break' statement before 'case'  no-fallthrough

✖ 1 problem (1 error, 0 warnings)
```

### Link to Minimal Reproducible Example

https://typescript-eslint.io/play#ts=5.0.4&sourceType=module&code=GYVwdgxgLglg9mABFAFgCgJQC5FgKYBueATogN4BQiyKxcA7ogKLF3GYDcFAvhRXgA8ADnGJREoSLAQS4cNARxgQAWwBGJbIgJwYAE3JVEAZ3owoEFIgUZD1ahACGxvIgCMWI-ZqcviJy6IAEye3g4IxnAANngAdFFwAOZoAEQA8gDSKRhcYYhqxHiOANa5iLzcQA&eslintrc=N4KABGBEBOD2sBdIC4wOgVwKYBpxQAcBDaAZy2hSgAEEBPArUgY2gEsCEBaJgGzYB2CAPTEyFSHggwMvJlVARpA2FwBmRXrwQALOBgDmOqpApxK+AL4hLQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA

### Participation

- [ ] I am willing to submit a pull request for this issue.
